doc/man/debos.1
'\" t
.\" Automatically generated by Pandoc 3.1.3
.\"
.\" Define V font for inline verbatim, using C font in formats
.\" that render this, and otherwise B font.
.ie "\f[CB]x\f[]"x" \{\
. ftr V B
. ftr VI BI
. ftr VB B
. ftr VBI BI
.\}
.el \{\
. ftr V CR
. ftr VI CI
. ftr VB CB
. ftr VBI CBI
.\}
.TH "debos" "1" "" "" ""
.hy
.SH NAME
.PP
debos - Debian OS images builder
.SH SYNOPSIS
.IP
.nf
\f[C]
debos [options] <recipe file in YAML>
debos [--help]
\f[R]
.fi
.PP
Application Options:
.IP
.nf
\f[C]
  -b, --fakemachine-backend=   Fakemachine backend to use (default: auto)
      --artifactdir=           Directory for packed archives and ostree repositories (default: current directory)
  -t, --template-var=          Template variables (use -t VARIABLE:VALUE syntax)
      --debug-shell            Fall into interactive shell on error
  -s, --shell=                 Redefine interactive shell binary (default: bash) (default: /bin/bash)
      --scratchsize=           Size of disk backed scratch space
  -c, --cpus=                  Number of CPUs to use for build VM (default: 2)
  -m, --memory=                Amount of memory for build VM (default: 2048MB)
      --show-boot              Show boot/console messages from the fake machine
  -e, --environ-var=           Environment variables (use -e VARIABLE:VALUE syntax)
  -v, --verbose                Verbose output
      --print-recipe           Print final recipe
      --dry-run                Compose final recipe to build but without any real work started
      --disable-fakemachine    Do not use fakemachine.
\f[R]
.fi
.SH DESCRIPTION
.PP
debos is a tool to make the creation of various Debian-based OS images
simpler.
While most other tools focus on specific use-cases, debos is designed to
be a toolchain making common actions trivial while providing enough rope
to do whatever tweaking which might be required behind the scenes.
.PP
debos expects a YAML file as input and runs the actions listed in the
file sequentially.
These actions should be self-contained and independent of each other.
.PP
Some of the actions provided by debos to customise and produce images
are:
.IP \[bu] 2
apt: install packages and their dependencies with `apt'
.IP \[bu] 2
debootstrap: construct the target rootfs with debootstrap
.IP \[bu] 2
download: download a single file from the internet
.IP \[bu] 2
filesystem-deploy: deploy a root filesystem to an image previously
created
.IP \[bu] 2
image-partition: create an image file, make partitions and format them
.IP \[bu] 2
ostree-commit: create an OSTree commit from rootfs
.IP \[bu] 2
ostree-deploy: deploy an OSTree branch to the image
.IP \[bu] 2
overlay: do a recursive copy of directories or files to the target
filesystem
.IP \[bu] 2
pack: create a tarball with the target filesystem
.IP \[bu] 2
pacman: install packages and their dependencies with pacman
.IP \[bu] 2
pacstrap: construct the target rootfs with pacstrap
.IP \[bu] 2
raw: directly write a file to the output image at a given offset
.IP \[bu] 2
recipe: includes the recipe actions at the given path
.IP \[bu] 2
run: allows to run a command or script in the filesystem or in the host
.IP \[bu] 2
unpack: unpack files from archive in the filesystem
.PP
A full syntax description of all the debos actions can be found at:
https://godoc.org/github.com/akaybayram61/debos-minimal/actions
.SH INSTALLATION (DOCKER CONTAINER)
.PP
An official debos container is available:
.IP
.nf
\f[C]
docker pull godebos/debos
\f[R]
.fi
.PP
See
docker/README.md (https://github.com/akaybayram61/debos-minimal/blob/master/docker/README.md)
for usage.
.SH INSTALLATION FROM SOURCE (UNDER DEBIAN)
.IP
.nf
\f[C]
sudo apt install golang git libglib2.0-dev libostree-dev qemu-system-x86 \[rs]
     qemu-user-static debootstrap systemd-container
export GOPATH=/opt/src/gocode # or whatever suits your needs
go install -v github.com/akaybayram61/debos-minimal/cmd/debos\[at]latest
/opt/src/gocode/bin/debos --help
\f[R]
.fi
.SH SIMPLE EXAMPLE
.PP
The following example will create an arm64 image, install several
packages in it, change the file /etc/hostname to \[lq]debian\[rq] and
finally make a tarball of the complete system.
.IP
.nf
\f[C]
{{- $image := or .image \[dq]debian.tgz\[dq] -}}

architecture: arm64

actions:
  - action: debootstrap
    suite: bookworm
    components:
      - main
      - non-free-firmware
    mirror: https://deb.debian.org/debian
    variant: minbase

  - action: apt
    packages: [ sudo, openssh-server, adduser, systemd-sysv, firmware-linux ]

  - action: run
    chroot: true
    command: echo debian > /etc/hostname

  - action: pack
    file: {{ $image }}
    compression: gz
\f[R]
.fi
.PP
To run it, create a file named \f[V]example.yaml\f[R] and run:
.IP
.nf
\f[C]
debos example.yaml
\f[R]
.fi
.PP
The final tarball will be named \[lq]debian.tgz\[rq] if you would like
to modify the fileame, you can provided a different name for the
variable image like this:
.IP
.nf
\f[C]
debos -t image:\[dq]debian-arm64.tgz\[dq] example.yaml
\f[R]
.fi
.SH OTHER EXAMPLES
.PP
Example recipes are collected in a separate repository:
.PP
https://github.com/akaybayram61/debos-minimal-recipes
.SH ENVIRONMENT VARIABLES
.PP
debos reads a predefined list of environment variables from the host and
propagates them to the fakemachine build environemnt.
The set of environment variables is defined by \f[V]environ_vars\f[R] in
\f[V]cmd/debos/debos.go\f[R].
Currently the list of environment variables includes the proxy
environment variables documented at:
.PP
https://wiki.archlinux.org/index.php/proxy_settings
.PP
The list of environment variables currently exported to fakemachine is:
.IP
.nf
\f[C]
http_proxy, https_proxy, ftp_proxy, rsync_proxy, all_proxy, no_proxy
\f[R]
.fi
.PP
While the elements of \f[V]environ_vars\f[R] are in lower case, for each
element both lower and upper case variants are probed on the host and if
found propagated to fakemachine.
So if the host has the environment variables HTTP_PROXY and no_proxy
defined, both will be propagated to fakemachine respecting the case.
.PP
The command line options \f[V]--environ-var\f[R] and \f[V]-e\f[R] can be
used to specify, overwrite and unset environment variables for
fakemachine with the syntax:
.IP
.nf
\f[C]
$ debos -e ENVIRONVAR:VALUE ...
\f[R]
.fi
.PP
To unset an environment variable, or in other words, to prevent an
environment variable being propagated to fakemachine, use the same
syntax without a value.
debos accepts multiple -e simultaneously.
.SH PROXY CONFIGURATION
.PP
While the proxy related environment variables are exported from the host
to fakemachine, there are two known sources of issues:
.IP \[bu] 2
Using localhost will not work from fakemachine.
Use an address which is valid on your network.
debos will warn if the environment variables contain localhost.
.IP \[bu] 2
In case you are running applications and/or scripts inside fakemachine
you may need to check which are the proxy environment variables they
use.
Different apps are known to use different environment variable names and
different case for environment variable names.
.SH FAKEMACHINE BACKEND
.PP
debos (unless running debos with the \f[V]--disable-fakemachine\f[R]
argument) creates and spawns a virtual machine using
fakemachine (https://github.com/go-debos/fakemachine) and executes the
actions defined by the recipe inside the virtual machine.
This helps ensure recipes are reproducible no matter the host
environment.
.PP
Fakemachine can use different virtualisation backends to spawn the
virtual machine, for more information see the fakemachine
documentation (https://github.com/go-debos/fakemachine).
.PP
By default the backend will automatically be selected based on what is
supported by the host machine, but this can be overridden using the
\f[V]--fakemachine-backend\f[R] / \f[V]-b\f[R] option.
If no backends are supported, debos reverts to running the recipe on the
host without creating a fakemachine.
.PP
Performance of the backends is roughly as follows: \f[V]kvm\f[R] is
faster than \f[V]uml\f[R] is faster than \f[V]qemu\f[R].
Using \f[V]--disable-fakemachine\f[R] is slightly faster than
\f[V]kvm\f[R], but requires root permissions.
.PP
Benchmark times for running
pine-a64-plus/debian.yaml (https://github.com/akaybayram61/debos-minimal-recipes/blob/9a25b4be6c9136f4a27e542f39ab7e419fc852c9/pine-a64-plus/debian.yaml)
on an Intel Pentium G4560T with SSD:
.PP
.TS
tab(@);
l l l.
T{
Backend
T}@T{
Wall Time
T}@T{
Prerequisites
T}
_
T{
\f[V]--disable-fakemachine\f[R]
T}@T{
8 min
T}@T{
root permissions
T}
T{
\f[V]-b kvm\f[R]
T}@T{
9 min
T}@T{
access to \f[V]/dev/kvm\f[R]
T}
T{
\f[V]-b uml\f[R]
T}@T{
18 min
T}@T{
package \f[V]user-mode-linux\f[R] installed
T}
T{
\f[V]-b qemu\f[R]
T}@T{
166 min
T}@T{
none
T}
.TE


cmd/debos/debos.go
package main

import (
	"fmt"
	"io/ioutil"
	"log"
	"os"
	"path"
	"strings"

	"github.com/akaybayram61/debos-minimal"
	"github.com/akaybayram61/debos-minimal/actions"
	"github.com/docker/go-units"
	"github.com/go-debos/fakemachine"
	"github.com/jessevdk/go-flags"
)

const Version = "1.0.4"

func checkError(context *debos.DebosContext, err error, a debos.Action, stage string) int {
	if err == nil {
		return 0
	}

	context.State = debos.Failed
	log.Printf("Action `%s` failed at stage %s, error: %s", a, stage, err)
	debos.DebugShell(*context)
	return 1
}

func do_run(r actions.Recipe, context *debos.DebosContext) int {
	for _, a := range r.Actions {
		log.Printf("==== %s ====\n", a)
		err := a.Run(context)

		// This does not stop the call of stacked Cleanup methods for other Actions
		// Stack Cleanup methods
		defer a.Cleanup(context)

		// Check the state of Run method
		if exitcode := checkError(context, err, a, "Run"); exitcode != 0 {
			return exitcode
		}
	}

	return 0
}

func warnLocalhost(variable string, value string) {
	message := `WARNING: Environment variable %[1]s contains a reference to
		    localhost. This may not work when running from fakemachine.
		    Consider using an address that is valid on your network.`

	if strings.Contains(value, "localhost") ||
		strings.Contains(value, "127.0.0.1") ||
		strings.Contains(value, "::1") {
		log.Printf(message, variable)
	}
}

func main() {
	context := debos.DebosContext{&debos.CommonContext{}, "", ""}
	var options struct {
		Backend            string            `short:"b" long:"fakemachine-backend" description:"Fakemachine backend to use" default:"auto"`
		ArtifactDir        string            `long:"artifactdir" description:"Directory for packed archives and ostree repositories (default: current directory)"`
		InternalImage      string            `long:"internal-image" hidden:"true"`
		TemplateVars       map[string]string `short:"t" long:"template-var" description:"Template variables (use -t VARIABLE:VALUE syntax)"`
		DebugShell         bool              `long:"debug-shell" description:"Fall into interactive shell on error"`
		Shell              string            `short:"s" long:"shell" description:"Redefine interactive shell binary (default: bash)" optionsl:"" default:"/bin/bash"`
		ScratchSize        string            `long:"scratchsize" description:"Size of disk backed scratch space"`
		CPUs               int               `short:"c" long:"cpus" description:"Number of CPUs to use for build VM (default: 2)"`
		Memory             string            `short:"m" long:"memory" description:"Amount of memory for build VM (default: 2048MB)"`
		ShowBoot           bool              `long:"show-boot" description:"Show boot/console messages from the fake machine"`
		EnvironVars        map[string]string `short:"e" long:"environ-var" description:"Environment variables (use -e VARIABLE:VALUE syntax)"`
		Verbose            bool              `short:"v" long:"verbose" description:"Verbose output"`
		PrintRecipe        bool              `long:"print-recipe" description:"Print final recipe"`
		DryRun             bool              `long:"dry-run" description:"Compose final recipe to build but without any real work started"`
		DisableFakeMachine bool              `long:"disable-fakemachine" description:"Do not use fakemachine."`
		Version            bool              `long:"version" description:"Print debos version"`
	}

	// These are the environment variables that will be detected on the
	// host and propagated to fakemachine. These are listed lower case, but
	// they are detected and configured in both lower case and upper case.
	var environ_vars = [...]string{
		"http_proxy",
		"https_proxy",
		"ftp_proxy",
		"rsync_proxy",
		"all_proxy",
		"no_proxy",
	}

	var exitcode int = 0
	// Allow to run all deferred calls prior to os.Exit()
	defer func() {
		os.Exit(exitcode)
	}()

	parser := flags.NewParser(&options, flags.Default)
	fakemachineBackends := parser.FindOptionByLongName("fakemachine-backend")
	fakemachineBackends.Choices = fakemachine.BackendNames()

	args, err := parser.Parse()
	if err != nil {
		flagsErr, ok := err.(*flags.Error)
		if ok && flagsErr.Type == flags.ErrHelp {
			return
		} else {
			exitcode = 1
			return
		}
	}

	if options.Version {
		fmt.Printf("debos-minimal v%s\n", Version)
		return
	}

	if len(args) != 1 {
		log.Println("No recipe given!")
		exitcode = 1
		return
	}

	if options.DisableFakeMachine && options.Backend != "auto" {
		log.Println("--disable-fakemachine and --fakemachine-backend are mutually exclusive")
		exitcode = 1
		return
	}

	// Set interactive shell binary only if '--debug-shell' options passed
	if options.DebugShell {
		context.DebugShell = options.Shell
	}

	if options.PrintRecipe {
		context.PrintRecipe = options.PrintRecipe
	}

	if options.Verbose {
		context.Verbose = options.Verbose
	}

	file := args[0]
	file = debos.CleanPath(file)

	r := actions.Recipe{}
	if _, err := os.Stat(file); os.IsNotExist(err) {
		log.Println(err)
		exitcode = 1
		return
	}
	if err := r.Parse(file, options.PrintRecipe, options.Verbose, options.TemplateVars); err != nil {
		log.Println(err)
		exitcode = 1
		return
	}

	/* If fakemachine is used the outer fake machine will never use the
	 * scratchdir, so just set it to /scratch as a dummy to prevent the
	 * outer debos creating a temporary directory */
	context.Scratchdir = "/scratch"

	var runInFakeMachine = true
	var m *fakemachine.Machine
	if options.DisableFakeMachine || fakemachine.InMachine() {
		runInFakeMachine = false
	} else {
		// attempt to create a fakemachine
		m, err = fakemachine.NewMachineWithBackend(options.Backend)
		if err != nil {
			log.Printf("error creating fakemachine: %v", err)

			/* fallback to running on the host unless the user has chosen
			 * a specific backend */
			if options.Backend == "auto" {
				runInFakeMachine = false
			} else {
				exitcode = 1
				return
			}
		}
	}

	// if running on the host create a scratchdir
	if !runInFakeMachine && !fakemachine.InMachine() {
		log.Printf("fakemachine not supported, running on the host!")
		cwd, _ := os.Getwd()
		context.Scratchdir, err = ioutil.TempDir(cwd, ".debos-")
		defer os.RemoveAll(context.Scratchdir)
	}

	context.Rootdir = path.Join(context.Scratchdir, "root")
	context.Image = options.InternalImage
	context.RecipeDir = path.Dir(file)

	context.Artifactdir = options.ArtifactDir
	if context.Artifactdir == "" {
		context.Artifactdir, _ = os.Getwd()
	}
	context.Artifactdir = debos.CleanPath(context.Artifactdir)

	// Initialise origins map
	context.Origins = make(map[string]string)
	context.Origins["artifacts"] = context.Artifactdir
	context.Origins["filesystem"] = context.Rootdir
	context.Origins["recipe"] = context.RecipeDir

	context.Architecture = r.Architecture

	context.State = debos.Success

	// Initialize environment variables map
	context.EnvironVars = make(map[string]string)

	// First add variables from host
	for _, e := range environ_vars {
		lowerVar := strings.ToLower(e) // lowercase not really needed
		lowerVal := os.Getenv(lowerVar)
		if lowerVal != "" {
			context.EnvironVars[lowerVar] = lowerVal
		}

		upperVar := strings.ToUpper(e)
		upperVal := os.Getenv(upperVar)
		if upperVal != "" {
			context.EnvironVars[upperVar] = upperVal
		}
	}

	// Then add/overwrite with variables from command line
	for k, v := range options.EnvironVars {
		// Allows the user to unset environ variables with -e
		if v == "" {
			delete(context.EnvironVars, k)
		} else {
			context.EnvironVars[k] = v
		}
	}

	for _, a := range r.Actions {
		err = a.Verify(&context)
		if exitcode = checkError(&context, err, a, "Verify"); exitcode != 0 {
			return
		}
	}

	if options.DryRun {
		log.Printf("==== Recipe done (Dry run) ====")
		return
	}

	if runInFakeMachine {
		var args []string

		if options.Memory == "" {
			// Set default memory size for fakemachine
			options.Memory = "2Gb"
		}
		memsize, err := units.RAMInBytes(options.Memory)
		if err != nil {
			fmt.Printf("Couldn't parse memory size: %v\n", err)
			exitcode = 1
			return
		}
		m.SetMemory(int(memsize / 1024 / 1024))

		if options.CPUs == 0 {
			// Set default CPU count for fakemachine
			options.CPUs = 2
		}
		m.SetNumCPUs(options.CPUs)

		if options.ScratchSize != "" {
			size, err := units.FromHumanSize(options.ScratchSize)
			if err != nil {
				fmt.Printf("Couldn't parse scratch size: %v\n", err)
				exitcode = 1
				return
			}
			m.SetScratch(size, "")
		}

		m.SetShowBoot(options.ShowBoot)

		// Puts in a format that is compatible with output of os.Environ()
		if context.EnvironVars != nil {
			EnvironString := []string{}
			for k, v := range context.EnvironVars {
				warnLocalhost(k, v)
				EnvironString = append(EnvironString, fmt.Sprintf("%s=%s", k, v))
			}
			m.SetEnviron(EnvironString) // And save the resulting environ vars on m
		}

		m.AddVolume(context.Artifactdir)
		args = append(args, "--artifactdir", context.Artifactdir)

		for k, v := range options.TemplateVars {
			args = append(args, "--template-var", fmt.Sprintf("%s:%s", k, v))
		}

		for k, v := range options.EnvironVars {
			args = append(args, "--environ-var", fmt.Sprintf("%s:%s", k, v))
		}

		m.AddVolume(context.RecipeDir)
		args = append(args, file)

		if options.DebugShell {
			args = append(args, "--debug-shell")
			args = append(args, "--shell", fmt.Sprintf("%s", options.Shell))
		}

		for _, a := range r.Actions {
			// Stack PostMachineCleanup methods
			defer a.PostMachineCleanup(&context)

			err = a.PreMachine(&context, m, &args)
			if exitcode = checkError(&context, err, a, "PreMachine"); exitcode != 0 {
				return
			}
		}

		// Silence extra output from fakemachine unless the --verbose flag was passed.
		m.SetQuiet(!options.Verbose)

		exitcode, err = m.RunInMachineWithArgs(args)
		if err != nil {
			fmt.Println(err)
			return
		}

		if exitcode != 0 {
			context.State = debos.Failed
			return
		}

		for _, a := range r.Actions {
			err = a.PostMachine(&context)
			if exitcode = checkError(&context, err, a, "Postmachine"); exitcode != 0 {
				return
			}
		}

		log.Printf("==== Recipe done ====")
		return
	}

	if !fakemachine.InMachine() {
		for _, a := range r.Actions {
			// Stack PostMachineCleanup methods
			defer a.PostMachineCleanup(&context)

			err = a.PreNoMachine(&context)
			if exitcode = checkError(&context, err, a, "PreNoMachine"); exitcode != 0 {
				return
			}
		}
	}

	// Create Rootdir
	if _, err = os.Stat(context.Rootdir); os.IsNotExist(err) {
		err = os.Mkdir(context.Rootdir, 0755)
		if err != nil && os.IsNotExist(err) {
			exitcode = 1
			return
		}
	}

	exitcode = do_run(r, &context)
	if exitcode != 0 {
		return
	}

	if !fakemachine.InMachine() {
		for _, a := range r.Actions {
			err = a.PostMachine(&context)
			if exitcode = checkError(&context, err, a, "PostMachine"); exitcode != 0 {
				return
			}
		}
		log.Printf("==== Recipe done ====")
	}
}
